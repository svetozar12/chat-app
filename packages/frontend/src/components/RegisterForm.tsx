import { useDispatch, useSelector } from "react-redux";
import Link from "next/dist/client/link";
import { InitialState, InitialState3 } from "../redux/state";

function RegisterForm({
  quickLogin,
  handleSubmit,
}: {
  quickLogin(): void;
  handleSubmit: (e: React.MouseEvent<HTMLButtonElement>) => Promise<void>;
}) {
  const dispatch = useDispatch();
  const state = useSelector(
    (state: { authReducer: InitialState }) => state.authReducer,
  );

  const inputState = useSelector(
    (state: { saveInputReducer: InitialState3 }) => state.saveInputReducer,
  );

  return (
    <div style={{ height: "100vh" }} className="register_form">
      <form className="container" style={{ height: "90vh" }}>
        <h2
          className="alert"
          style={state.good ? { color: "green" } : { color: "red" }}
        >
          {state.good || state.bad}
        </h2>
        <h1>Register</h1>
        <input
          value={inputState.input_username}
          onChange={(e) =>
            dispatch({ type: "SAVE_INPUT_USERNAME", payload: e.target.value })
          }
          type="text"
          name="username"
          placeholder="username ..."
        />
        <input
          value={inputState.input_password}
          onChange={(e) =>
            dispatch({ type: "SAVE_INPUT_PASSWORD", payload: e.target.value })
          }
          type="password"
          name="password"
          placeholder="password ..."
        />
        <input
          value={inputState.input_email}
          onChange={(e) =>
            dispatch({ type: "SAVE_INPUT_EMAIL", payload: e.target.value })
          }
          type="email"
          name="email"
          placeholder="email ..."
        />
        <div className="input_gender">
          <span>
            <label htmlFor="Male">Male</label>
            <input
              onChange={(e) =>
                dispatch({
                  type: "SAVE_INPUT_GENDER",
                  payload: e.target.value,
                })
              }
              type="radio"
              name="gender"
              id="Male"
              value="Male"
            />
          </span>
          <span>
            <label htmlFor="Female">Female</label>
            <input
              onChange={(e) =>
                dispatch({
                  type: "SAVE_INPUT_GENDER",
                  payload: e.target.value,
                })
              }
              type="radio"
              name="gender"
              id="Female"
              value="Female"
            />
          </span>
          <span>
            <label htmlFor="Others">Others</label>
            <input
              onChange={(e) =>
                dispatch({
                  type: "SAVE_INPUT_GENDER",
                  payload: e.target.value,
                })
              }
              type="radio"
              name="gender"
              id="Others"
              value="Others"
            />
          </span>
        </div>
        <button onClick={handleSubmit} type="submit">
          Register
        </button>
        <Link href="/login">
          <a className="link" style={{ color: "var(--main-blue)" }}>
            Already have an account?
          </a>
        </Link>
        {state.loginPrompt && (
          <div
            onClick={quickLogin}
            style={{ cursor: "pointer" }}
            className="quick_login"
          >
            <h2>Click me to Quick login</h2>
          </div>
        )}
      </form>
    </div>
  );
}

export default RegisterForm;
