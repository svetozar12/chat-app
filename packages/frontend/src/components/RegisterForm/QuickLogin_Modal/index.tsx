export { default } from "./QuickLogin_Modal";
