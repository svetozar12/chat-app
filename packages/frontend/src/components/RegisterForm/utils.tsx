import { IFields } from 'components/FormWrapper/FormWrapper';
import { Gender } from 'services/generated';

interface IValues {
  username: string;
  email: string;
  password: string;
  gender: Gender;
}

export const renderInputs = (color: string): IFields[] => [
  {
    props: {
      type: 'text',
      name: 'username',
      placeholder: 'username ...',
      boxShadow: `0px 0px 2px 0px ${color}`,
      _placeholder: { color, opacity: 0.5 },
    },
  },
  {
    props: {
      type: 'password',
      name: 'password',
      placeholder: 'password ...',
      boxShadow: `0px 0px 2px 0px ${color}`,
      _placeholder: { color, opacity: 0.5 },
    },
  },
  {
    props: {
      type: 'email',
      name: 'email',
      placeholder: 'email ...',
      boxShadow: `0px 0px 2px 0px ${color}`,
      _placeholder: { color, opacity: 0.5 },
    },
  },
  {
    props: { type: 'radio' },
    radioButtons: [
      {
        value: Gender.Male,
        props: { name: 'gender', id: Gender.Male },
      },
      { value: Gender.Female, props: { name: 'gender', id: Gender.Female } },
    ],
  },
];
