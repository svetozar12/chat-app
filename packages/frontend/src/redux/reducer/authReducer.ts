import { Action } from "../state";
import { ActionType } from "../state";
import { InitialState } from "../state";
import { HYDRATE } from "next-redux-wrapper";

const initialState = {
  remember_me: false,
  input: "",
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: "",
};
const reducer = (state: InitialState = initialState, action: Action) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.SAVE_INPUT:
      return { ...state, input: action.payload };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    case ActionType.SIGN_OUT:
      return { state: undefined };
    default:
      return state;
  }
};

export default reducer;
