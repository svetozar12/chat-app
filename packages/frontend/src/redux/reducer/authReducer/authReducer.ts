import { HYDRATE } from "next-redux-wrapper";
import { IAuthState } from "./state";
import { ActionType } from "../../types";
export const initialState: IAuthState = {
  remember_me: false,
  good: "", //good and bad stand for good alert and bad alert
  bad: "",
  loginPrompt: false,
  cookie: {
    username: "",
    id: "",
    token: "",
    refresh_token: "",
  },
};
const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.LOGIN_POST:
      return { ...state, state };
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.good };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.bad };
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.QUICK_LOGIN:
      return { ...state, loginPrompt: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default reducer;
export { initialState as AuthState };
