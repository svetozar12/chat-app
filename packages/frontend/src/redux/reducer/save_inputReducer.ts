import { Action, ActionType, InitialState3 } from "../state";
import { HYDRATE } from "next-redux-wrapper";
export const initialState = {
  input_username: "",
  input_password: "",
  input_email: "",
  input_gender: "",
  notification_number: 0,
};
const saveInputReducer = (state: InitialState3 = initialState, action: Action) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.SAVE_INPUT_USERNAME:
      return { ...state, input_username: action.payload };
    case ActionType.SAVE_INPUT_PASSWORD:
      return { ...state, input_password: action.payload };
    case ActionType.SAVE_INPUT_EMAIL:
      return { ...state, input_email: action.payload };
    case ActionType.SAVE_INPUT_GENDER:
      return { ...state, input_gender: action.payload };
    case ActionType.NOTIFICATION_NUMBER:
      return { ...state, notification_number: action.payload };
    default:
      return state;
  }
};

export default saveInputReducer;
