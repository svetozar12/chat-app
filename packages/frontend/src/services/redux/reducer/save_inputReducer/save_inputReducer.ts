import ISave_inputState from "./state";
import { HYDRATE } from "next-redux-wrapper";
import { ActionType } from "../../types";

export const initialState: ISave_inputState | ISave_inputState[] = {
  input_username: "",
  input_password: "",
  input_email: "",
  input_gender: "other",
  input_file: null,
  notification_number: 0,
};
const saveInputReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case HYDRATE:
      return {
        ...state,
        ...action.payload,
      };
    case ActionType.SAVE_INPUT_USERNAME:
      return { ...state, input_username: action.payload };
    case ActionType.SAVE_INPUT_PASSWORD:
      return { ...state, input_password: action.payload };
    case ActionType.SAVE_INPUT_EMAIL:
      return { ...state, input_email: action.payload };
    case ActionType.SAVE_INPUT_GENDER:
      return { ...state, input_gender: action.payload };
    case ActionType.SAVE_INPUT_FILE:
      return { ...state, input_file: action.payload };
    case ActionType.NOTIFICATION_NUMBER:
      return { ...state, notification_number: action.payload };
    default:
      return state;
  }
};

export default saveInputReducer;
