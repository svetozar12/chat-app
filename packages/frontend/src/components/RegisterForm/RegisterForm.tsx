import { connect } from 'react-redux';
// services
import { Flex, Heading } from '@chakra-ui/react';
import React, { FC } from 'react';
import DefaultLink from '../../services/chat-ui/DefaultLink';
// components
import QuickLogin from './QuickLogin';
import FormWrapper from 'components/FormWrapper';
import useThemeColors from '../../hooks/useThemeColors';
import { toggleQuickLogin } from 'services/redux/reducer/toggles/actions';
import { bindActionCreators, Dispatch } from 'redux';
import { STATE } from 'services/redux/reducer';
import { IToggle } from 'services/redux/reducer/toggles/state';
import { Gender, useCreateUserMutation } from 'services/generated';
import { setAlert } from 'services/redux/reducer/alert/actions';
import { renderInputs } from 'components/RegisterForm/utils';

interface IRegisterForm extends ReturnType<typeof mapDispatchToProps> {
  toggle: IToggle;
}
type FormValues = { username: string; password: string; email: string; gender: Gender };
const RegisterForm: FC<IRegisterForm> = ({ setAlert, toggle, toggleQuickLogin }) => {
  const {
    base: {
      button: { color: btnColor },
      default: { color },
    },
  } = useThemeColors();
  const [createUserMutation, { data }] = useCreateUserMutation();
  const { createUser } = data || {};
  if (createUser?.__typename === 'Error') {
    setAlert(createUser.message, 'error');
    return <></>;
  }

  const handleSubmit = async ({ username, password, email, gender }: FormValues) => {
    if (!username || !password || !email || !gender) return;
    await createUserMutation({ variables: { user: { username, email, password, gender } } });
    const { createUser } = data || {};
    if (createUser?.__typename === 'Error') setAlert(createUser.message, 'error');
    else {
      setAlert('User created', 'success');
      toggleQuickLogin(true);
    }
  };

  return (
    <>
      <FormWrapper
        onSubmit={handleSubmit}
        header={
          <Flex w="full" justifyContent="center" gap={2} mb="2" flexDir="column">
            <Heading>Register</Heading>
            <DefaultLink href="/" text="Already have an account?" />
          </Flex>
        }
        fields={renderInputs(color)}
        buttons={[{ value: 'Register', props: { colorScheme: btnColor, w: '30%', type: 'submit' } }]}
      />
      {toggle.toggleQuickLogin && (
        <QuickLogin
          UserId={createUser?.userId as string}
          AccessToken={createUser?.AccessToken as string}
          RefreshToken={createUser?.RefreshToken as string}
        />
      )}
    </>
  );
};

const mapStateToProps = (state: STATE) => ({
  toggle: state.toggle,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  toggleQuickLogin: bindActionCreators(toggleQuickLogin, dispatch),
  setAlert: bindActionCreators(setAlert, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(RegisterForm);
