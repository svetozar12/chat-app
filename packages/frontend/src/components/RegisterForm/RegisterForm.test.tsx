import RegisterForm from "../RegisterForm";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { initialState } from "../../redux/reducer/authReducer/authReducer";
import { initialState as saveInputState } from "../../redux/reducer/save_inputReducer/save_inputReducer";
import configureStore from "redux-mock-store";
const quickLogin: any = jest.fn();
const handleSubmit = jest.fn();

const mockStore = configureStore([]);
const store = mockStore({
  authReducer: initialState,
  saveInputReducer: saveInputState,
});

const setupRender = () => {
  const component = render(
    <Provider store={store}>
      <RegisterForm quickLogin={quickLogin} isLogging={false} handleSubmit={handleSubmit} />
    </Provider>,
  );
  return component;
};

describe("Render connected React-redux page", () => {
  beforeEach(() => {
    setupRender();
  });
  it("should create snapshot for <RegisterForm/>", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <RegisterForm quickLogin={quickLogin} isLogging={false} handleSubmit={handleSubmit} />
          </Provider>,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });

  it("should render <RegisterForm/>", () => {
    const renderedComponent = screen.getByTitle("register_form");
    expect(renderedComponent).toBeInTheDocument();
  });
});
