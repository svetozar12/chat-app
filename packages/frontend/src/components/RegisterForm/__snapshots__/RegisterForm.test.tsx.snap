// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render connected React-redux page should create snapshot for <RegisterForm/> 1`] = `
<div
  className="flex"
  style={
    Object {
      "flexDirection": "column",
      "height": "100vh",
    }
  }
  title="register_form"
>
  
  <h1
    className="css-my1y1y"
  >
    Register
  </h1>
  <form
    className="css-7rx0xo"
    style={
      Object {
        "height": "70vh",
      }
    }
  >
    <div
      className="css-artkr0"
    >
      <label>
        Username
      </label>
      <input
        className="css-1s11o3x"
        name="username"
        onChange={[Function]}
        placeholder="username ..."
        type="text"
        value=""
      />
    </div>
    <div
      className="css-artkr0"
    >
      <label>
        Password
      </label>
      <input
        className="css-1s11o3x"
        name="password"
        onChange={[Function]}
        placeholder="password ..."
        type="password"
        value=""
      />
    </div>
    <div
      className="css-artkr0"
    >
      <label>
        Email
      </label>
      <input
        className="css-1s11o3x"
        name="email"
        onChange={[Function]}
        placeholder="email ..."
        type="email"
        value=""
      />
    </div>
    <div
      className="css-artkr0"
    >
      <label>
        Gender
      </label>
      <div
        className="flex"
      >
        <input
          className="css-alxow4 css-1s11o3x"
          id="Male"
          name="gender"
          onChange={[Function]}
          type="radio"
          value="Male"
        />
        <label
          className="css-pa4pqz"
          htmlFor="Male"
        >
          Male
        </label>
        <div>
          <input
            className="css-19gixz3 css-1s11o3x"
            id="Female"
            name="gender"
            onChange={[Function]}
            type="radio"
            value="Female"
          />
          <label
            className="css-pa4pqz"
            htmlFor="Female"
          >
            Female
          </label>
        </div>
      </div>
    </div>
    <button
      className="css-1hcgzlw"
      onClick={[MockFunction]}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      style={
        Object {
          "color": "var(--button-blue)",
        }
      }
    >
      Already have an account?
    </a>
  </form>
</div>
`;
